<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-condition differential expression analysis in single-cell RNA-seq data using Poisson mash • poisson.mash.alpha</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multi-condition differential expression analysis in single-cell RNA-seq data using Poisson mash">
<meta property="og:description" content="poisson.mash.alpha">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poisson.mash.alpha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1-93</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Multi-condition differential expression analysis
in single-cell RNA-seq data using Poisson mash</h1>
                        <h4 data-toc-skip class="author">Yusha Liu</h4>
            
            <h4 data-toc-skip class="date">2023-11-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/poisson.mash.alpha/blob/HEAD/vignettes/intro_poisson_mash.Rmd" class="external-link"><code>vignettes/intro_poisson_mash.Rmd</code></a></small>
      <div class="hidden name"><code>intro_poisson_mash.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>This vignette illustrates how to use <a href="https://arxiv.org/abs/2210.00697" class="external-link">Poisson multivariate adaptive
shrinkage</a> to analyze a data set. Poisson mash was developed for
multi-condition differential expression analysis with single-cell
RNA-sequencing (scRNA-seq) data in which gene expression is measured in
many (e.g., dozens) of different conditions. The aim is to detect which
genes are differentially expressed (DE) and to estimate expression
differences (log-fold changes) among multiple conditions. This approach
makes use of the multivariate adaptive shrinkage (<a href="https://dx.doi.org/10.1038/s41588-018-0268-8" class="external-link">mash</a>) prior.</p>
<p>We illustrate how to implement this approach by analyzing data from
an experiment intended to study the effects of cytokine stimulation on
gene expression in neutrophils collected from cytokine-injected
mice.</p>
<p>We first load in the R packages that will be used in the analysis,
then we work through an example to illustrate how to apply Poisson mash
step-by-step.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willtownes/glmpca" class="external-link">glmpca</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">poisson.mash.alpha</a></span><span class="op">)</span></code></pre></div>
<p>Also, set the random number generator seed to ensure that the results
are more reproducible.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-a-data-object-for-poisson-mash-analysis">Create a data object for Poisson mash analysis<a class="anchor" aria-label="anchor" href="#create-a-data-object-for-poisson-mash-analysis"></a>
</h3>
<p>To run Poisson mash, you need to provide (1) a matrix of UMI counts,
stored as a <span class="math inline">\(J \times N\)</span> matrix <span class="math inline">\(Y\)</span> with entries <span class="math inline">\(y_{ji}\)</span>, where <span class="math inline">\(i = 1, \dots,N\)</span> indexes cells and <span class="math inline">\(j=1,\dots,J\)</span> indexes genes; and (2) an
vector of length <span class="math inline">\(n\)</span> in which the
<span class="math inline">\(i^{\mathrm{th}}\)</span> element gives the
condition label of cell <span class="math inline">\(i\)</span>. We
assume that the <span class="math inline">\(N\)</span> cells come from
<span class="math inline">\(R\)</span> conditions (in practice, <span class="math inline">\(R\)</span> should be considerably larger than
2).</p>
<p>The original neutrophils dataset contained counts for 13,362 cells,
8,543 genes and 45 conditions (44 cytokine treatments plus one control).
For the purposes of illustrating the interface, we use a small subset of
these data (1,936 cells, 2,158 genes, 12 conditions), but later we
examine the results from running Poisson mash on the full data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">neutrophils_subset</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span>
<span class="co"># [1] 2158 1936</span>
<span class="co"># conditions</span>
<span class="co">#    IL10 IL12p70    IL18    IL1a    IL1b     IL2     IL3     IL4     IL5     IL6 </span>
<span class="co">#      82     198     254     392     251     109     105     133      81     122 </span>
<span class="co">#     IL7     IL9 </span>
<span class="co">#     105     104</span></code></pre></div>
<p>After loading in the data, we compute the cell-specific size factors.
This can be done in various ways. The simplest option is to take the sum
of the counts <span class="math inline">\(y_{ji}\)</span> across all
genes in each cell <span class="math inline">\(i\)</span>. Here we take
a deconvolution-based approach that was implemented in the scran
package:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
<span class="va">si</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">calculateSumFactors</a></span><span class="op">(</span><span class="va">Y</span>,clusters <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">si</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></code></pre></div>
<p>Next, we create a data object for poisson mash analysis using the
function <code>pois_mash_set_data</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash_set_data.html">pois_mash_set_data</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">conditions</span>,<span class="va">si</span><span class="op">)</span></code></pre></div>
<p>By default, Poisson mash assumes a gene-specific baseline expression
level, <span class="math inline">\(\mu_j\)</span>, shared across all
conditions. Poisson mash can also handle the more complex setting in
which the <span class="math inline">\(R\)</span> conditions arise from
<span class="math inline">\(m = 1, \dots, M\)</span> subgroups, and each
subgroup has its own baseline, <span class="math inline">\(\mu_{jm}\)</span>. For example, one may want to
perform a DE analysis across multiple treatment conditions, jointly for
multiple cell types. In this case, <span class="math inline">\(M\)</span> is the number of cell types, and <span class="math inline">\(R\)</span> represents the total number of
combinations of treatment conditions and cell types. Please refer to the
documentation for the function <code>pois_mash_set_data</code> to see
how to do this.</p>
</div>
<div class="section level3">
<h3 id="estimate-latent-factors-capturing-unwanted-variation">Estimate latent factors capturing unwanted variation<a class="anchor" aria-label="anchor" href="#estimate-latent-factors-capturing-unwanted-variation"></a>
</h3>
<p>Unwanted variation present in the data can induce dependence among
gene-wise tests and confound DE analysis. In some cases, the confounding
variables are known, such as when these capture batch effects. More
often, the unwanted variation is due to unmeasured factors. Here we
estimating the confounding variables using <a href="https://github.com/willtownes/glmpca" class="external-link">glmpca</a>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">conditions</span><span class="op">)</span>
<span class="va">fit.glmpca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/glmpca/man/glmpca.html" class="external-link">glmpca</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>,X <span class="op">=</span> <span class="va">design</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>,L  <span class="op">=</span> <span class="fl">4</span>,fam <span class="op">=</span> <span class="st">"nb2"</span>,
                     sz <span class="op">=</span> <span class="va">si</span>,ctl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>minIter <span class="op">=</span> <span class="fl">1</span>,maxIter <span class="op">=</span> <span class="fl">20</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Warning in glmpca(Y = as.matrix(Y), X = design[, -1], L = 4, fam = "nb2", :</span>
<span class="co"># Reached maximum number of iterations (20) without numerical convergence. Results</span>
<span class="co"># may be unreliable.</span>
<span class="va">Fuv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">fit.glmpca</span><span class="op">$</span><span class="va">loadings</span><span class="op">)</span></code></pre></div>
<p>To save time, we ran <code>glmpca</code> for only 20 iterations, but
typically more iterations will be needed (100, or perhaps more).</p>
</div>
<div class="section level3">
<h3 id="fit-the-poisson-mash-model">Fit the Poisson mash model<a class="anchor" aria-label="anchor" href="#fit-the-poisson-mash-model"></a>
</h3>
<div class="section level4">
<h4 id="step-1-prefit-the-model-to-initialize-parameters">Step 1: Prefit the model to initialize parameters<a class="anchor" aria-label="anchor" href="#step-1-prefit-the-model-to-initialize-parameters"></a>
</h4>
<p>We first fit Poisson mash under the assumption of no DE to get
sensible initial estimates of the other model parameters, including the
gene-specific baseline expressions <span class="math inline">\(\mu_j\)</span>. Please refer to the <a href="https://arxiv.org/abs/2210.00697" class="external-link">paper</a> for details.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prefit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash_prefit.html">pois_mash_prefit</a></span><span class="op">(</span><span class="va">dat</span>,ruv <span class="op">=</span> <span class="cn">TRUE</span>,Fuv <span class="op">=</span> <span class="va">Fuv</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co"># Initializing psi2 via grid search.</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2-set-up-the-prior-covariance-matrices">Step 2: Set up the prior covariance matrices<a class="anchor" aria-label="anchor" href="#step-2-set-up-the-prior-covariance-matrices"></a>
</h4>
<p>Let the <span class="math inline">\(R \times 1\)</span> vector <span class="math inline">\(\beta_j = (\beta_{j1}, \dots,
\beta_{jR})'\)</span> denote the true DE effects of each of the
<span class="math inline">\(R\)</span> conditions relative to the
baseline expression <span class="math inline">\(\mu_j\)</span> for gene
<span class="math inline">\(j\)</span>. We assume that <span class="math inline">\(\beta_j\)</span> follows a mash prior, which is a
mixture of multivariate normal distributions: <span class="math display">\[ p(\beta_j; \pi, U) =
\sum_{k=1}^K \sum_{l=1}^L \pi_{kl} N_R(\beta_j; 0, w_l U_k).\]</span> In
the mash prior, <span class="math inline">\(w_l\)</span> is a
pre-specified grid of scaling coefficients, spanning from very small to
very large, to capture the full range of possible DE effect sizes; <span class="math inline">\(U_k\)</span> is a set of covariance matrices, each
of which capturing a different pattern of covariation of DE effects
across conditions; <span class="math inline">\(\pi_{kl}\)</span> are
corresponding mixture proportions that are estimated from the data
during model fitting.</p>
<p>To use Poisson mash, you need to specify the prior covariance
matrices. This maay include “canonical” covariance matrices that
represent, for example, DE specific to a condition; and “data-driven”
covariance matrices that are learned from the data and can capture
arbitrary patterns of DE among conditions.</p>
<p>For the canonical matrices, here we consider the matrices capturing
condition-specific effects for each of the <span class="math inline">\(R\)</span> conditions. These matrices can be
generate using the function <code>pois_cov_canonical</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ulist.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_canonical.html">pois_cov_canonical</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></code></pre></div>
<p>Data-driven covariance matrices are estimated using a three-step
process: (1) initialization, (2) refinement, (3) diagonal entry
modification. The initialization phase is implemented by the function
<code>pois_cov_init</code>. To estimate data-driven covariance matrices,
we only use the subset of genes containing “strong signals”; that is,
the genes that are considered DE by a multinomial goodness-of-fit
test.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_init.html">pois_cov_init</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></code></pre></div>
<p>The refinement phase refines the initial estimates of data-driven
covariance matrices using the Extreme Eeconvolution (ED) algorithm, and
is implemented by the function <code>pois_cov_ed</code>. As in the
initialization phase, only data from the genes with strong signals are
used. For the purposes of this vignette, we just ran
<code>pois_cov_ed</code> for 10 iterations, but we recommend using the
default value <code>maxiter = 500</code> when analyzing your own
data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">R</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">X</span><span class="op">)</span>
<span class="va">ulist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res.pca</span><span class="op">$</span><span class="va">ulist</span>,<span class="va">ulist.c</span><span class="op">)</span>
<span class="co"># Specify whether each prior covariance matrix is data-driven or not.</span>
<span class="va">ulist.dd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">TRUE</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">res.pca</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">R</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> 
<span class="co"># Run ED on the genes with strong signals.</span>
<span class="va">fit.ed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_ed.html">pois_cov_ed</a></span><span class="op">(</span><span class="va">dat</span>,subset <span class="op">=</span> <span class="va">res.pca</span><span class="op">$</span><span class="va">subset</span>,Ulist <span class="op">=</span> <span class="va">res.pca</span><span class="op">$</span><span class="va">Ulist</span>,
                      ulist <span class="op">=</span> <span class="va">ulist</span>,ulist.dd <span class="op">=</span> <span class="va">ulist.dd</span>,ruv <span class="op">=</span> <span class="cn">TRUE</span>,
                      Fuv <span class="op">=</span> <span class="va">Fuv</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span>,init <span class="op">=</span> <span class="va">prefit</span>,
                      control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxiter <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Start running extreme deconvolution to estimate prior covariance matrices.</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "1:    -24608.666465"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "2:    -24345.614492"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "3:    -24153.516130"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "4:    -23985.986148"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "5:    -23857.513161"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "6:    -23760.482292"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "7:    -23682.316394"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "8:    -23611.507470"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "9:    -23541.062859"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "10:    -23470.835015"</span>
<span class="co"># Finish running extreme deconvolution to estimate prior covariance matrices.</span></code></pre></div>
<p>The modification phase adds a small constant (e.g., <span class="math inline">\(0.01\)</span>) to the diagonal entries of the
estimated data-driven covariance matrices. This step is done to
alleviate issues of inflated <a href="https://doi.org/10.1093/biostatistics/kxw041" class="external-link">local false sign
rates (lfsr)</a>, which are analogous to but more conservative than
local false discovery rates.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># add epsilon*I to each full-rank data-driven prior covariance matrix</span>
<span class="va">Ulist</span> <span class="op">&lt;-</span> <span class="va">fit.ed</span><span class="op">$</span><span class="va">Ulist</span>
<span class="va">H</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Ulist</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">h</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">H</span><span class="op">)</span><span class="op">{</span>
  <span class="va">Uh</span> <span class="op">&lt;-</span> <span class="va">Ulist</span><span class="op">[[</span><span class="va">h</span><span class="op">]</span><span class="op">]</span>
  <span class="va">Uh</span> <span class="op">&lt;-</span> <span class="va">Uh</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Uh</span><span class="op">)</span><span class="op">)</span>
  <span class="va">Ulist</span><span class="op">[[</span><span class="va">h</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">Uh</span> <span class="op">+</span> <span class="fl">1e-2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">R</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># add epsilon*I to each rank-1 data-driven prior covariance matrix</span>
<span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span>
<span class="va">epsilon2.G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1e-8</span>, <span class="va">G</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">epsilon2.G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span>
<span class="va">epsilon2.G</span><span class="op">[</span><span class="va">ulist.dd</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1e-2</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-3-fit-the-poisson-mash-model">Step 3: Fit the Poisson mash model<a class="anchor" aria-label="anchor" href="#step-3-fit-the-poisson-mash-model"></a>
</h4>
<p>Now we are ready to fit the Poisson mash model to data from all
genes, which is implemented by the function <code>pois_mash</code>. For
the purposes of this illustration, we only run <code>pois_mash</code>
for 10 iterations, but we in practice recommend setting
<code>maxiter = 100</code> (or more) when analyzing your own data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash.html">pois_mash</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>,Ulist <span class="op">=</span> <span class="va">Ulist</span>,ulist <span class="op">=</span> <span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span>,
                 ulist.epsilon2 <span class="op">=</span> <span class="va">epsilon2.G</span>,gridmult <span class="op">=</span> <span class="fl">2.5</span>,ruv <span class="op">=</span> <span class="cn">TRUE</span>,
                 Fuv <span class="op">=</span> <span class="va">Fuv</span>,rho <span class="op">=</span> <span class="va">prefit</span><span class="op">$</span><span class="va">rho</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span>,
                 init <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="va">prefit</span><span class="op">$</span><span class="va">mu</span>,psi2 <span class="op">=</span> <span class="va">prefit</span><span class="op">$</span><span class="va">psi2</span><span class="op">)</span>,
                 control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxiter <span class="op">=</span> <span class="fl">10</span>,nc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> 
<span class="co"># Start fitting Poisson mash model.</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "1:    -67705.176984"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "1:    667"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "2:    -67431.051782"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "2:    463"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "3:    -67279.622606"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "3:    460"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "4:    -67154.400780"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "4:    558"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "5:    -67053.393501"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "5:    634"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "6:    -66970.902547"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "6:    655"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "7:    -66903.055342"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "7:    704"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "8:    -66846.947201"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "8:    722"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "9:    -66800.982448"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "9:    720"</span>
<span class="co"># [1] "iter         ELBO"</span>
<span class="co"># [1] "10:    -66763.296734"</span>
<span class="co"># [1] "iter         number_of_j_to_update"</span>
<span class="co"># [1] "10:    737"</span>
<span class="co"># Finished fitting Poisson mash model.</span>
<span class="co"># Start calculating posterior summary.</span>
<span class="co"># Finish calculating posterior summary.</span></code></pre></div>
<p>Poisson mash aims to detect and estimate expression differences
relative to some “reference” condition, which needs to be specified by
the user. By default, the reference condition is the mean across all
conditions. Alternatively, users can specify a particular condition as
the reference condition using the arguments <code>C</code> and
<code>res.colnames</code> of the function <code>pois_mash</code>. If it
is also of interest to detect and estimate expression differences
relative to the median across all conditions, this can be accomplished
by setting <code>median_deviations = TRUE</code> in the above call.</p>
</div>
</div>
<div class="section level2">
<h2 id="examining-the-results">Examining the results<a class="anchor" aria-label="anchor" href="#examining-the-results"></a>
</h2>
<p>We will next look at the results, which are stored in the
<code>res</code> object produced by the call to <code>pois_mash</code>.
Recall that we illustrate how to fit the Poisson mash model using a
subset of the neutrophils data to save time. Here we load in the Poisson
mash fit to the entire neutrophils dataset.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">)</span></code></pre></div>
<p>The output contains posterior summaries of the condition-specific DE
effects relative to the reference condition specified by the user,
including posterior means and posterior standard deviations of the DE
effect sizes (log-fold changes) and measures of significance (lfsr), for
each gene in the dataset. These DE effect posterior summaries are stored
in the <code>result</code> component.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Rransform DE effect estimates to the scale of log2 fold change.</span>
<span class="va">res</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">PosteriorMean</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>
<span class="co"># gene-specific, condition-specific lfsr</span>
<span class="va">res</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">lfsr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#          Ctrl_2-mean   CCL20-mean    CXCL1-mean    CCL22-mean   CXCL5-mean</span>
<span class="co"># Mrpl15  -0.034042592 -0.007468183  0.0104375255 -0.0212980437 -0.017696220</span>
<span class="co"># Lypla1  -0.009057545 -0.002816239 -0.0312773830 -0.0100671020 -0.011641034</span>
<span class="co"># Tcea1    0.006406279  0.003163708  0.0117792404  0.0112956807  0.007417445</span>
<span class="co"># Atp6v1h  0.005641873  0.004324441 -0.0009984417  0.0001755186  0.002388297</span>
<span class="co"># Rb1cc1   0.005531303  0.010590542 -0.0071790326  0.0060939489  0.003342450</span>
<span class="co">#         Ctrl_2-mean CCL20-mean CXCL1-mean CCL22-mean CXCL5-mean</span>
<span class="co"># Mrpl15    0.2527081  0.4012463  0.4501408  0.3236194  0.3458751</span>
<span class="co"># Lypla1    0.4041496  0.4586630  0.2824110  0.3962331  0.3803850</span>
<span class="co"># Tcea1     0.4273815  0.4727043  0.3991172  0.3839887  0.4144509</span>
<span class="co"># Atp6v1h   0.4512667  0.4597141  0.4917574  0.4894398  0.4710016</span>
<span class="co"># Rb1cc1    0.4402427  0.3958559  0.4170690  0.4337283  0.4628668</span></code></pre></div>
<p>We can find the number of DE genes, which are defined as genes having
<em>lfsr</em> less than <span class="math inline">\(\alpha\)</span> in
at least one condition, where <span class="math inline">\(\alpha\)</span> is a threshold specified by the
user (e.g., 0.05).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfsr</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">lfsr</span>
<span class="co"># Get the indices of identified DE genes.</span>
<span class="va">min.lfsr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">lfsr</span>,<span class="fl">1</span>,<span class="va">min</span><span class="op">)</span>
<span class="co"># Order genes by significance based on mininum lfsr across conditions.</span>
<span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">min.lfsr</span><span class="op">)</span>
<span class="va">lfsr</span> <span class="op">&lt;-</span> <span class="va">lfsr</span><span class="op">[</span><span class="va">rows</span>,<span class="op">]</span>
<span class="va">min.lfsr</span> <span class="op">&lt;-</span> <span class="va">min.lfsr</span><span class="op">[</span><span class="va">rows</span><span class="op">]</span>
<span class="va">de.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">min.lfsr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co"># Get the total number of DE genes.</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">de.genes</span><span class="op">)</span>
<span class="co"># [1] 2636</span></code></pre></div>
<p>The model parameter estimates are stored in the
<code>pois.mash.fit</code> component. In particular, let us look at the
estimates of the mixture proportions for different prior covariance
matrices:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">pi</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span>
<span class="co">#    tPCA Emp_cov    PC_1    PC_2    PC_3    PC_4    PC_5     u_0     e_1     e_2 </span>
<span class="co">#   0.912   0.002   0.002   0.010   0.034   0.012   0.004   0.001   0.000   0.001 </span>
<span class="co">#     e_3     e_4     e_5     e_6     e_7     e_8     e_9    e_10    e_11    e_12 </span>
<span class="co">#   0.000   0.000   0.000   0.000   0.001   0.000   0.000   0.001   0.000   0.000 </span>
<span class="co">#    e_13    e_14    e_15    e_16    e_17    e_18    e_19    e_20    e_21    e_22 </span>
<span class="co">#   0.000   0.001   0.000   0.000   0.000   0.000   0.000   0.000   0.000   0.000 </span>
<span class="co">#    e_23    e_24    e_25    e_26    e_27    e_28    e_29    e_30    e_31    e_32 </span>
<span class="co">#   0.000   0.000   0.000   0.001   0.000   0.000   0.000   0.000   0.000   0.000 </span>
<span class="co">#    e_33    e_34    e_35    e_36    e_37    e_38    e_39    e_40    e_41    e_42 </span>
<span class="co">#   0.000   0.004   0.001   0.000   0.000   0.000   0.000   0.000   0.002   0.001 </span>
<span class="co">#    e_43    e_44    e_45 </span>
<span class="co">#   0.000   0.000   0.000</span></code></pre></div>
<p>The data-driven covariance matrix “tPCA” has by far the largest
proportion estimate (91%). We plot the correlation heatmap for this
covariance matrix, which captures the most frequent pattern of sharing
of DE effects across cytokine treatments in the neutrophils data. For
space reason, here we present the correlation heatmap only for selected
conditions.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">condition_colors</span><span class="op">)</span>
<span class="va">U</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">Ulist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">U</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">U</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span>
<span class="va">idx.trts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols.trt</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span>
<span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cov2cor</a></span><span class="op">(</span><span class="va">U</span><span class="op">[</span><span class="va">idx.trts</span>,<span class="va">idx.trts</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">corr</span>,cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,angle_col <span class="op">=</span> <span class="fl">90</span>,
         fontsize <span class="op">=</span> <span class="fl">12</span>,main <span class="op">=</span> <span class="st">"Major prior covariance matrix (weight = 91%)"</span><span class="op">)</span></code></pre></div>
<p><img src="intro_poisson_mash_files/figure-html/plot-sharing-pattern-1.png" width="600" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># R version 3.6.2 (2019-12-12)</span>
<span class="co"># Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co"># Running under: macOS Catalina 10.15.7</span>
<span class="co"># </span>
<span class="co"># Matrix products: default</span>
<span class="co"># BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co"># LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co"># </span>
<span class="co"># locale:</span>
<span class="co"># [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co"># </span>
<span class="co"># attached base packages:</span>
<span class="co"># [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co"># [8] methods   base     </span>
<span class="co"># </span>
<span class="co"># other attached packages:</span>
<span class="co">#  [1] poisson.mash.alpha_0.1-92   pheatmap_1.0.12            </span>
<span class="co">#  [3] glmpca_0.2.0                scran_1.14.6               </span>
<span class="co">#  [5] SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.1</span>
<span class="co">#  [7] DelayedArray_0.12.3         BiocParallel_1.18.1        </span>
<span class="co">#  [9] matrixStats_0.63.0          Biobase_2.46.0             </span>
<span class="co"># [11] GenomicRanges_1.38.0        GenomeInfoDb_1.22.1        </span>
<span class="co"># [13] IRanges_2.20.2              S4Vectors_0.24.4           </span>
<span class="co"># [15] BiocGenerics_0.32.0         Matrix_1.3-4               </span>
<span class="co"># </span>
<span class="co"># loaded via a namespace (and not attached):</span>
<span class="co">#  [1] bitops_1.0-6             fs_1.5.2                 RColorBrewer_1.1-2      </span>
<span class="co">#  [4] rprojroot_2.0.3          tools_3.6.2              bslib_0.3.1             </span>
<span class="co">#  [7] utf8_1.1.4               R6_2.4.1                 irlba_2.3.3             </span>
<span class="co"># [10] vipor_0.4.5              DBI_1.1.0                colorspace_1.4-1        </span>
<span class="co"># [13] tidyselect_1.1.1         gridExtra_2.3            compiler_3.6.2          </span>
<span class="co"># [16] cli_3.5.0                BiocNeighbors_1.2.0      desc_1.2.0              </span>
<span class="co"># [19] sass_0.4.0               scales_1.1.0             mvtnorm_1.0-11          </span>
<span class="co"># [22] SQUAREM_2017.10-1        mixsqp_0.3-48            pkgdown_2.0.7           </span>
<span class="co"># [25] systemfonts_1.0.2        stringr_1.4.0            digest_0.6.23           </span>
<span class="co"># [28] rmarkdown_2.21           XVector_0.26.0           scater_1.14.6           </span>
<span class="co"># [31] pkgconfig_2.0.3          htmltools_0.5.4          highr_0.8               </span>
<span class="co"># [34] invgamma_1.1             fastmap_1.1.0            limma_3.42.2            </span>
<span class="co"># [37] rlang_1.0.6              DelayedMatrixStats_1.6.1 jquerylib_0.1.4         </span>
<span class="co"># [40] generics_0.0.2           jsonlite_1.7.2           dplyr_1.0.7             </span>
<span class="co"># [43] RCurl_1.98-1.2           magrittr_2.0.1           BiocSingular_1.2.2      </span>
<span class="co"># [46] GenomeInfoDbData_1.2.2   Rcpp_1.0.8               ggbeeswarm_0.6.0        </span>
<span class="co"># [49] munsell_0.5.0            fansi_0.4.0              abind_1.4-5             </span>
<span class="co"># [52] viridis_0.5.1            lifecycle_1.0.3          stringi_1.4.3           </span>
<span class="co"># [55] yaml_2.2.0               edgeR_3.28.1             MASS_7.3-51.4           </span>
<span class="co"># [58] zlibbioc_1.32.0          grid_3.6.2               dqrng_0.2.1             </span>
<span class="co"># [61] crayon_1.4.1             lattice_0.20-38          locfit_1.5-9.4          </span>
<span class="co"># [64] knitr_1.37               pillar_1.6.2             igraph_1.2.5            </span>
<span class="co"># [67] glue_1.4.2               evaluate_0.14            vctrs_0.3.8             </span>
<span class="co"># [70] gtable_0.3.0             purrr_0.3.4              assertthat_0.2.1        </span>
<span class="co"># [73] ashr_2.2-57              ggplot2_3.3.6            xfun_0.36               </span>
<span class="co"># [76] rsvd_1.0.2               ragg_0.3.1               viridisLite_0.3.0       </span>
<span class="co"># [79] truncnorm_1.0-8          tibble_3.1.3             poilog_0.4              </span>
<span class="co"># [82] beeswarm_0.2.3           memoise_1.1.0            seqgendiff_1.2.2        </span>
<span class="co"># [85] statmod_1.4.34           ellipsis_0.3.2</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yusha Liu, Peter Carbonetto, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
