<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-condition differential expression analysis in single-cell RNA-seq data using Poisson mash • poisson.mash.alpha</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multi-condition differential expression analysis in single-cell RNA-seq data using Poisson mash">
<meta property="og:description" content="poisson.mash.alpha">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poisson.mash.alpha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1-88</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Multi-condition differential expression analysis
in single-cell RNA-seq data using Poisson mash</h1>
                        <h4 data-toc-skip class="author">Yusha Liu</h4>
            
            <h4 data-toc-skip class="date">2023-11-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/poisson.mash.alpha/blob/HEAD/vignettes/intro.Rmd" class="external-link"><code>vignettes/intro.Rmd</code></a></small>
      <div class="hidden name"><code>intro.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>This vignette shows how to apply “Poisson multivariate adaptive
shrinkage”, or Poisson mash, from <a href="https://arxiv.org/abs/2210.00697" class="external-link">Liu et al.</a>, using the <a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">poisson.mash.alpha</a>
software. Poisson mash was developed to perform multi-condition
differential expression analyses on single cell RNA-sequencing
(scRNA-seq) data, in which gene expression is measured in many (e.g.,
dozens) different conditions and the aim is to detect which genes are
differentially expressed (DE) and to estimate expression differences
(log-fold changes) among multiple conditions. This approach makes use of
the multivariate adaptive shrinkage (<a href="https://dx.doi.org/10.1038/s41588-018-0268-8" class="external-link">mash</a>) prior
introduced previously by our lab.</p>
<p>We illustrate how to implement this approach by analyzing data from
an <a href="https://arxiv.org/abs/2210.00697" class="external-link">experiment</a> conducted
to study the effects of cytokine stimulation on gene expression in
neutrophils collected from cytokine-injected mice.</p>
<p>We first load in the R packages that will be used in the analysis,
then go through an example to illustrate how to apply Poisson mash step
by step.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eweine/glmpca" class="external-link">glmpca</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">poisson.mash.alpha</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-a-data-object-for-poisson-mash-analysis">Create a data object for Poisson mash analysis<a class="anchor" aria-label="anchor" href="#create-a-data-object-for-poisson-mash-analysis"></a>
</h3>
<p>To run Poisson mash, users need to provide (1) a matrix of UMI
counts, which is stored as a <span class="math inline">\(J \times
N\)</span> matrix <span class="math inline">\(Y\)</span> with entries
<span class="math inline">\(y_{ji}\)</span>, where <span class="math inline">\(i=1,\dots,N\)</span> indexes cells and <span class="math inline">\(j=1,\dots,J\)</span> indexes genes, and (2) a
<span class="math inline">\(N \times 1\)</span> vector with its <span class="math inline">\(i^{\text{th}}\)</span> element indicating the
condition label of cell <span class="math inline">\(i\)</span>. We
assume that the <span class="math inline">\(N\)</span> cells come from
<span class="math inline">\(R\)</span> conditions and focus on
multi-condition experiments with <span class="math inline">\(R \gg
2\)</span>. As a concrete example, the original neutrophils dataset
contains counts from 13,362 cells in 8,543 genes from 45 conditions (44
cytokine treatments and one control). For time reason, here we show how
to fit Poisson mash using a subset of the dataset containing 1,936 cells
in 2,158 genes from 12 conditions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"data/neutrophils_subset.RData"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span>
<span class="va">R</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>After loading in the data, we compute the cell-specific size factors,
which can be done in various ways. The simplest option is to take the
sum of the counts <span class="math inline">\(y_{ji}\)</span> across all
genes in each cell <span class="math inline">\(i\)</span>, but here we
take a more robust deconvolution-based approach.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
<span class="va">si</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">calculateSumFactors</a></span><span class="op">(</span><span class="va">Y</span>, clusters<span class="op">=</span><span class="va">clusters</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">si</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></code></pre></div>
<p>Next, we create a data object for poisson mash analysis using the
function <code>pois_mash_set_data</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash_set_data.html">pois_mash_set_data</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">conditions</span>, <span class="va">si</span><span class="op">)</span></code></pre></div>
<p>By default, Poisson mash assumes a gene-specific baseline expression
level, <span class="math inline">\(\mu_j\)</span>, which is shared
across all conditions. Poisson mash can also handle the more complex
setting in which the <span class="math inline">\(R\)</span> conditions
arise from <span class="math inline">\(m = 1, \dots, M\)</span>
subgroups, and each subgroup has its own baseline, <span class="math inline">\(\mu_{jm}\)</span>. For example, one may want to
perform a DE analysis across multiple treatment conditions, jointly for
multiple cell types. In this case, <span class="math inline">\(M\)</span> is the number of cell types, and <span class="math inline">\(R\)</span> represents the total number of
combinations of treatment conditions and cell types. Please refer to the
documentation for the function <code>pois_mash_set_data</code> to see
how to do this.</p>
</div>
<div class="section level3">
<h3 id="estimate-the-latent-factors-causing-unwanted-variation">Estimate the latent factors causing unwanted variation<a class="anchor" aria-label="anchor" href="#estimate-the-latent-factors-causing-unwanted-variation"></a>
</h3>
<p>Unwanted variation present in the data can induce dependence among
gene-wise tests and confound DE analysis. In some cases, the confounding
variables are known, such as when these capture batch effects. More
often, the unwanted variation is due to unmeasured factors. Poisson mash
accounts for unwanted variation by estimating the latent confounding
factors using the R package <a href="https://github.com/willtownes/glmpca" class="external-link">glmpca</a>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">conditions</span><span class="op">)</span>
<span class="va">fit.glmpca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/glmpca/man/glmpca.html" class="external-link">glmpca</a></span><span class="op">(</span>Y<span class="op">=</span><span class="va">Y</span>, X<span class="op">=</span><span class="va">design</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>, L<span class="op">=</span><span class="fl">4</span>, fam<span class="op">=</span><span class="st">"nb2"</span>, sz<span class="op">=</span><span class="va">si</span>, ctl<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>verbose<span class="op">=</span><span class="cn">FALSE</span>, maxIter<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span>
<span class="va">Fuv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">fit.glmpca</span><span class="op">$</span><span class="va">loadings</span><span class="op">)</span>
<span class="va">runtime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">start_time</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">runtime</span><span class="op">)</span></code></pre></div>
<p>To save time, we just run <code>glmpca</code> for 100 iterations, but
we recommend setting <code>maxIter = 500</code> for better estimation of
latent confounding factors.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="fit-the-model">Fit the model<a class="anchor" aria-label="anchor" href="#fit-the-model"></a>
</h2>
<div class="section level3">
<h3 class="unnumbered" id="step-1--prefit-the-model-to-initialize-parameters">Step 1. Prefit the model to initialize
parameters<a class="anchor" aria-label="anchor" href="#step-1--prefit-the-model-to-initialize-parameters"></a>
</h3>
<p>We first fit Poisson mash under the assumption of no DE to get
sensible initial estimates of the other model parameters, including the
gene-specific baseline expression <span class="math inline">\(\mu_j\)</span>. Please refer to the model
description sections in our <a href="https://arxiv.org/abs/2210.00697" class="external-link">paper</a> for more details.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prefit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash_prefit.html">pois_mash_prefit</a></span><span class="op">(</span><span class="va">data</span>, ruv<span class="op">=</span><span class="cn">TRUE</span>, Fuv<span class="op">=</span><span class="va">Fuv</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 class="unnumbered" id="step-2--set-up-the-prior-covariance-matrices">Step 2. Set up the prior covariance matrices<a class="anchor" aria-label="anchor" href="#step-2--set-up-the-prior-covariance-matrices"></a>
</h3>
<p>Let the <span class="math inline">\(R \times 1\)</span> vector <span class="math inline">\(\beta_j = (\beta_{j1}, \dots,
\beta_{jR})'\)</span> denote the true DE effects of each of the
<span class="math inline">\(R\)</span> conditions relative to the
baseline expression <span class="math inline">\(\mu_j\)</span> for gene
<span class="math inline">\(j\)</span>. We assume that <span class="math inline">\(\beta_j\)</span> follows a mash prior, which is a
mixture of multivariate normal distributions: <span class="math display">\[ p(\beta_j; \pi, U) = \sum_{k=1}^K \sum_{l=1}^L
\pi_{kl} N_R(\beta_j; 0, w_l U_k).\]</span> In the mash prior, <span class="math inline">\(w_l\)</span> is a pre-specified grid of scaling
coefficients, spanning from very small to very large, to capture the
full range of possible DE effect sizes; <span class="math inline">\(U_k\)</span> is a set of covariance matrices, each
of which capturing a different pattern of covariation of DE effects
across conditions; <span class="math inline">\(\pi_{kl}\)</span> are
corresponding mixture proportions that are estimated from the data
during model fitting.</p>
<p>To use Poisson mash, users need to specify prior covariance matrices,
which include both “canonical” covariance matrices that represent, for
example, DE specific to a condition, and “data-driven” covariance
matrices that are learned from the data and can capture arbitrary
patterns of DE among conditions.</p>
<p>For the canonical matrices, here we consider the matrices capturing
condition-specific effects for each of the <span class="math inline">\(R\)</span> conditions. These matrices can be
defined using the function <code>pois_cov_canonical</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ulist.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_canonical.html">pois_cov_canonical</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<p>The data-driven covariance matrices are estimated using a three-step
process: (1) initialization, (2) refinement, (3) diagonal entry
modification. The initialization phase is implemented by the function
<code>pois_cov_init</code>. To estimate data-driven covariance matrices,
we only use the subset of genes containing “strong signals”, i.e., the
genes that are considered DE by a multinomial goodness-of-fit test.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_init.html">pois_cov_init</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<p>The refinement phase refines the initial estimates of data-driven
covariance matrices using an extreme deconvolution algorithm, and is
implemented by the function <code>pois_cov_ed</code>. As in the
initialization phase, only data from the genes with strong signals are
used. To save time, below we just run <code>pois_cov_ed</code> for 100
iterations, but we recommend using the default value
<code>maxiter = 500</code> when analyzing your own data.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ulist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res.pca</span><span class="op">$</span><span class="va">ulist</span>, <span class="va">ulist.c</span><span class="op">)</span>

<span class="co"># specify whether each prior covariance matrix is data-driven or not</span>
<span class="va">ulist.dd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">res.pca</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="va">R</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> 

<span class="co"># run extreme deconvolution on the genes with strong signals</span>
<span class="va">start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">fit.ed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_ed.html">pois_cov_ed</a></span><span class="op">(</span><span class="va">data</span>, subset<span class="op">=</span><span class="va">res.pca</span><span class="op">$</span><span class="va">subset</span>, Ulist<span class="op">=</span><span class="va">res.pca</span><span class="op">$</span><span class="va">Ulist</span>, ulist<span class="op">=</span><span class="va">ulist</span>, ulist.dd<span class="op">=</span><span class="va">ulist.dd</span>,
                      ruv<span class="op">=</span><span class="cn">TRUE</span>, Fuv<span class="op">=</span><span class="va">Fuv</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, init<span class="op">=</span><span class="va">prefit</span>, control<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxiter<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span>
<span class="va">runtime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">start_time</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">runtime</span><span class="op">)</span></code></pre></div>
<p>The modification phase adds a small constant (e.g., <span class="math inline">\(0.01\)</span>) to the diagonal entries of the
estimated data-driven covariance matrices. This step is done to
alleviate issues of inflated <a href="https://doi.org/10.1093/biostatistics/kxw041" class="external-link">local false sign
rates (lfsr)</a>, which are analogous to but more conservative than
local false discovery rates.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># add epsilon*I to each full-rank data-driven prior covariance matrix</span>
<span class="va">Ulist</span> <span class="op">&lt;-</span> <span class="va">fit.ed</span><span class="op">$</span><span class="va">Ulist</span>
<span class="va">H</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Ulist</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">h</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">H</span><span class="op">)</span><span class="op">{</span>
  <span class="va">Uh</span> <span class="op">&lt;-</span> <span class="va">Ulist</span><span class="op">[[</span><span class="va">h</span><span class="op">]</span><span class="op">]</span>
  <span class="va">Uh</span> <span class="op">&lt;-</span> <span class="va">Uh</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Uh</span><span class="op">)</span><span class="op">)</span>
  <span class="va">Ulist</span><span class="op">[[</span><span class="va">h</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">Uh</span> <span class="op">+</span> <span class="fl">1e-2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">R</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># add epsilon*I to each rank-1 data-driven prior covariance matrix</span>
<span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span>
<span class="va">epsilon2.G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1e-8</span>, <span class="va">G</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">epsilon2.G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span>
<span class="va">epsilon2.G</span><span class="op">[</span><span class="va">ulist.dd</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1e-2</span></code></pre></div>
</div>
<div class="section level3">
<h3 class="unnumbered" id="step-3--fit-the-poisson-mash-model">Step 3. Fit the Poisson mash model<a class="anchor" aria-label="anchor" href="#step-3--fit-the-poisson-mash-model"></a>
</h3>
<p>Now we are ready to fit the Poisson mash model to data from all
genes, which is implemented by the function <code>pois_mash</code>. To
save time, we just run <code>pois_mash</code> for 5 iterations, but we
recommend setting at least <code>maxiter = 100</code> when analyzing
your own data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">start_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash.html">pois_mash</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">data</span>, Ulist<span class="op">=</span><span class="va">Ulist</span>, ulist<span class="op">=</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span>, ulist.epsilon2<span class="op">=</span><span class="va">epsilon2.G</span>, gridmult<span class="op">=</span><span class="fl">2.5</span>, ruv<span class="op">=</span><span class="cn">TRUE</span>, Fuv<span class="op">=</span><span class="va">Fuv</span>, rho<span class="op">=</span><span class="va">prefit</span><span class="op">$</span><span class="va">rho</span>, 
                 verbose<span class="op">=</span><span class="cn">TRUE</span>, init<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mu<span class="op">=</span><span class="va">prefit</span><span class="op">$</span><span class="va">mu</span>, psi2<span class="op">=</span><span class="va">prefit</span><span class="op">$</span><span class="va">psi2</span><span class="op">)</span>, control<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxiter<span class="op">=</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span> 
<span class="va">runtime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">start_time</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">runtime</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">res</span>, file <span class="op">=</span> <span class="st">"data/neutrophils_subset_pois_mash_ruv_fit.rds"</span><span class="op">)</span></code></pre></div>
<p>Of note, Poisson mash aims to detect and estimate expression
differences relative to some “reference” condition, which needs to be
specified by the user. By default, the reference condition is the mean
across all conditions. Alternatively, users can specify a particular
condition as the reference condition using the arguments <code>C</code>
and <code>res.colnames</code> of the function <code>pois_mash</code>. If
it is also of interest to detect and estimate expression differences
relative to the median across all conditions, this can be accomplished
by setting <code>median_deviations = TRUE</code> in the above call.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="understand-the-output">Understand the output<a class="anchor" aria-label="anchor" href="#understand-the-output"></a>
</h2>
<p>We will next look at the results, which are stored in the
<code>res</code> object produced by the call to <code>pois_mash</code>.
Recall that we illustrate how to fit the Poisson mash model using a
subset of the neutrophils data to save time. Here we load in the Poisson
mash fit to the entire neutrophils dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"data/neutrophils_pois_mash_ruv_fit.RData"</span><span class="op">)</span></code></pre></div>
<p>The output contains posterior summaries of the condition-specific DE
effects relative to the reference condition specified by the user,
including posterior means and posterior standard deviations of the DE
effect sizes (log-fold changes) and measures of significance (lfsr), for
each gene in the dataset. These DE effect posterior summaries are stored
in the <code>result</code> component.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># transform DE effect estimates to the scale of log2 fold change</span>
<span class="va">res</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">PosteriorMean</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>
<span class="co"># gene-specific, condition-specific lfsr</span>
<span class="va">res</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">lfsr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<p>We can find the number of DE genes, which are defined as genes having
lfsr less than <span class="math inline">\(\alpha\)</span> in at least
one condition, where <span class="math inline">\(\alpha\)</span> is a
threshold specified by the user (e.g., <span class="math inline">\(0.05\)</span>).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfsr</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">lfsr</span>
<span class="co"># get the indices of identified DE genes</span>
<span class="va">de.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">lfsr</span>, <span class="fl">1</span>, <span class="va">min</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co"># get the total number of DE genes</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">de.genes</span><span class="op">)</span>
<span class="co"># order DE genes by significance based on mininum lfsr across conditions</span>
<span class="va">de.genes</span> <span class="op">&lt;-</span> <span class="va">de.genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">lfsr</span><span class="op">[</span><span class="va">de.genes</span>,<span class="op">]</span>, <span class="fl">1</span>, <span class="va">min</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>    
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">de.genes</span><span class="op">)</span></code></pre></div>
<p>The model parameter estimates are stored in the
<code>pois.mash.fit</code> component. In particular, let us look at the
estimates of the mixture proportions for different prior covariance
matrices:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">pi</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p>The data-driven covariance matrix “tPCA” has by far the largest
proportion estimate (<span class="math inline">\(91\%\)</span>). We plot
the correlation heatmap for this covariance matrix, which captures the
most frequent pattern of sharing of DE effects across cytokine
treatments in the neutrophils data. For space reason, here we present
the correlation heatmap only for selected conditions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"data/condition_colors.RData"</span><span class="op">)</span>
<span class="va">U</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">Ulist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">U</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">U</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span>
<span class="va">idx.trts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols.trt</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span>
<span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cov2cor</a></span><span class="op">(</span><span class="va">U</span><span class="op">[</span><span class="va">idx.trts</span>, <span class="va">idx.trts</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">corr</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, angle_col <span class="op">=</span> <span class="fl">90</span>, fontsize <span class="op">=</span> <span class="fl">15</span>, main <span class="op">=</span> <span class="st">"Major prior covariance matrix (weight = 91%)"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yusha Liu, Peter Carbonetto, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
