<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-condition differential expression analysis in single-cell RNA-seq data using Poisson mash • poisson.mash.alpha</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multi-condition differential expression analysis in single-cell RNA-seq data using Poisson mash">
<meta property="og:description" content="poisson.mash.alpha">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poisson.mash.alpha</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1-100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Multi-condition differential expression analysis
in single-cell RNA-seq data using Poisson mash</h1>
                        <h4 data-toc-skip class="author">Yusha Liu,
Peter Carbonetto</h4>
            
            <h4 data-toc-skip class="date">2023-11-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/poisson.mash.alpha/blob/HEAD/vignettes/intro_poisson_mash.Rmd" class="external-link"><code>vignettes/intro_poisson_mash.Rmd</code></a></small>
      <div class="hidden name"><code>intro_poisson_mash.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>This vignette illustrates how to use <a href="https://arxiv.org/abs/2210.00697" class="external-link">Poisson multivariate adaptive
shrinkage</a> to analyze a data set. Poisson mash was developed for
multi-condition differential expression analysis with single-cell
RNA-sequencing (scRNA-seq) data in which gene expression is measured in
many (e.g., dozens) of different conditions. The aim is to detect which
genes are differentially expressed (DE) and to estimate expression
differences (log-fold changes) among multiple conditions. This approach
makes use of the multivariate adaptive shrinkage (<a href="https://dx.doi.org/10.1038/s41588-018-0268-8" class="external-link">mash</a>) prior.</p>
<p>We illustrate this approach by analyzing scRNA-seq data for learning
about the effects of cytokine stimulation on gene expression. We will
work through the Poisson mash analysis step-by-step.</p>
<p>First, we load the R packages needed to run the analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willtownes/glmpca" class="external-link">glmpca</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/poisson.mash.alpha" class="external-link">poisson.mash.alpha</a></span><span class="op">)</span></code></pre></div>
<p>We also set the random number generator seed to ensure that the
results are reproducible.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-a-data-object-for-poisson-mash-analysis">Create a data object for Poisson mash analysis<a class="anchor" aria-label="anchor" href="#create-a-data-object-for-poisson-mash-analysis"></a>
</h3>
<p>To run Poisson mash, you need to provide: (1) a matrix of UMI counts,
stored as a <span class="math inline">\(J \times N\)</span> matrix <span class="math inline">\(Y\)</span> with entries <span class="math inline">\(y_{ji}\)</span>, where <span class="math inline">\(i = 1, \dots,N\)</span> indexes cells and <span class="math inline">\(j=1,\dots,J\)</span> indexes genes; and (2) a
vector of length <span class="math inline">\(n\)</span> in which the
<span class="math inline">\(i^{\mathrm{th}}\)</span> element gives the
condition label of cell <span class="math inline">\(i\)</span>. We
denote the number of conditions by <span class="math inline">\(R\)</span>. (<span class="math inline">\(R\)</span> should be much larger than 2.)</p>
<p>The original data contained counts for 13,362 cells, 8,543 genes and
45 conditions (44 cytokine treatments plus one control). For the this
vignette, we will use subset (1,936 cells, 2,158 genes, 12 conditions),
but later we will inspect the results obtained from running Poisson mash
on the full data set.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">neutrophils_subset</span><span class="op">)</span>
<span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">neutrophils_subset</span><span class="op">$</span><span class="va">Y</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="va">neutrophils_subset</span><span class="op">$</span><span class="va">conditions</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
<span class="va">conditions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">conditions</span><span class="op">)</span>
<span class="co"># [1] 2158 1936</span>
<span class="co"># conditions</span>
<span class="co">#    IL10 IL12p70    IL18    IL1a    IL1b     IL2     IL3     IL4     IL5     IL6 </span>
<span class="co">#      82     198     254     392     251     109     105     133      81     122 </span>
<span class="co">#     IL7     IL9 </span>
<span class="co">#     105     104</span></code></pre></div>
<p>After loading in the data, we compute the cell-specific size factors.
This can be done in various ways. The simplest option is to take the sum
of the counts <span class="math inline">\(y_{ji}\)</span> across all
genes in each cell <span class="math inline">\(i\)</span>. Here we take
a deconvolution-based approach implemented in the <strong>scran</strong>
package:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
<span class="va">si</span>        <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">calculateSumFactors</a></span><span class="op">(</span><span class="va">Y</span>,clusters <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">si</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></code></pre></div>
<p>Next, we create a data object for poisson mash analysis using the
function <code>pois_mash_set_data</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash_set_data.html">pois_mash_set_data</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">conditions</span>,<span class="va">si</span><span class="op">)</span></code></pre></div>
<p>By default, Poisson mash assumes a gene-specific baseline expression
level, <span class="math inline">\(\mu_j\)</span>, that is shared across
all conditions.</p>
<p>(Poisson mash can also handle the more complex setting in which the
<span class="math inline">\(R\)</span> conditions arise from <span class="math inline">\(m = 1, \dots, M\)</span> subgroups, and each
subgroup has its own baseline, <span class="math inline">\(\mu_{jm}\)</span>. For example, one may want to
perform a DE analysis across multiple treatment conditions, jointly for
multiple cell types. In this case, <span class="math inline">\(M\)</span> is the number of cell types, and <span class="math inline">\(R\)</span> represents the total number of
combinations of treatment conditions and cell types. Please refer to the
documentation for the function <code>pois_mash_set_data</code> to see
how to do this.)</p>
</div>
<div class="section level3">
<h3 id="estimate-latent-factors-capturing-unwanted-variation">Estimate latent factors capturing unwanted variation<a class="anchor" aria-label="anchor" href="#estimate-latent-factors-capturing-unwanted-variation"></a>
</h3>
<p>Unwanted variation present in the data can induce dependence among
gene-wise tests and confound the DE analysis. In some cases, the
confounding variables are known, such as when these capture batch
effects. More often, the unwanted variation is due to unmeasured
factors. Here we estimating the confounding variables using <a href="https://github.com/willtownes/glmpca" class="external-link">glmpca</a>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">conditions</span><span class="op">)</span>
<span class="va">fit.glmpca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/glmpca/man/glmpca.html" class="external-link">glmpca</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>,X <span class="op">=</span> <span class="va">design</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span>,L  <span class="op">=</span> <span class="fl">4</span>,fam <span class="op">=</span> <span class="st">"nb2"</span>,
                     sz <span class="op">=</span> <span class="va">si</span>,ctl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>minIter <span class="op">=</span> <span class="fl">1</span>,maxIter <span class="op">=</span> <span class="fl">20</span>,
                                        verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="va">Fuv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">fit.glmpca</span><span class="op">$</span><span class="va">loadings</span><span class="op">)</span></code></pre></div>
<p>For this vignette, we ran <code>glmpca</code> for only 20 iterations,
but typically it is better to perform more iterations (100, or perhaps
more).</p>
</div>
<div class="section level3">
<h3 id="fit-the-poisson-mash-model">Fit the Poisson mash model<a class="anchor" aria-label="anchor" href="#fit-the-poisson-mash-model"></a>
</h3>
<div class="section level4">
<h4 id="step-1-prefit-the-model">Step 1: Prefit the model<a class="anchor" aria-label="anchor" href="#step-1-prefit-the-model"></a>
</h4>
<p>We first fit Poisson mash under the assumption of no DE to get
sensible initial estimates of the other model parameters, including the
gene-specific baseline expressions <span class="math inline">\(\mu_j\)</span>. (See the <a href="https://arxiv.org/abs/2210.00697" class="external-link">paper</a> for details.)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prefit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash_prefit.html">pois_mash_prefit</a></span><span class="op">(</span><span class="va">dat</span>,ruv <span class="op">=</span> <span class="cn">TRUE</span>,Fuv <span class="op">=</span> <span class="va">Fuv</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2-set-up-the-prior-covariance-matrices">Step 2: Set up the prior covariance matrices<a class="anchor" aria-label="anchor" href="#step-2-set-up-the-prior-covariance-matrices"></a>
</h4>
<p>Let the <span class="math inline">\(R \times 1\)</span> vector <span class="math inline">\(\beta_j = (\beta_{j1}, \dots,
\beta_{jR})'\)</span> denote the true DE effects of each of the
<span class="math inline">\(R\)</span> conditions relative to the
baseline expression <span class="math inline">\(\mu_j\)</span> for gene
<span class="math inline">\(j\)</span>. We assume that <span class="math inline">\(\beta_j\)</span> follows a “mash prior”, which is
a mixture of multivariate normal distributions: <span class="math display">\[
p(\beta_j; \pi, U) =
\sum_{k=1}^K \sum_{l=1}^L \pi_{kl} N_R(\beta_j; 0, w_l U_k).
\]</span> In the mash prior, <span class="math inline">\(w_l\)</span> is
a pre-specified grid of scaling coefficients, spanning from very small
to very large to capture the full range of possible DE effect sizes;
<span class="math inline">\(U_k\)</span> is a set of covariance
matrices, each of which capture a different pattern of covariation
across conditions; <span class="math inline">\(\pi_{kl}\)</span> are
mixture proportions that are estimated from the data during model
fitting.</p>
<p>To use Poisson mash, you need to specify these prior covariance
matrices. This may include “canonical” covariance matrices that
represent, for example, DE specific to a condition; and “data-driven”
covariance matrices that are learned from the data and can capture
arbitrary patterns of DE among conditions.</p>
<p>For the canonical matrices, here we consider the matrices capturing
condition-specific effects for each of the <span class="math inline">\(R\)</span> conditions. These matrices can be
generated using the <code>pois_cov_canonical</code> function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ulist.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_canonical.html">pois_cov_canonical</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></code></pre></div>
<p>Data-driven covariance matrices are estimated using a three-step
process: (1) initialization, (2) refinement, (3) diagonal entry
modification. The initialization phase is implemented by the function
<code>pois_cov_init</code>. To estimate data-driven covariance matrices,
we only use the subset of genes containing “strong signals”; that is,
the genes that are considered DE by a multinomial goodness-of-fit test.
The selection of genes with strong signals is implicitly performed by
the call to <code>pois_cov_init</code>, e.g., setting
<code>cutoff = 3</code> selects the genes for which the magnitude of
z-score (inverted from the p-value of the goodness-of-fit test) exceeds
3.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_init.html">pois_cov_init</a></span><span class="op">(</span><span class="va">dat</span>,cutoff <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p>The refinement phase refines the initial estimates of data-driven
covariance matrices using the Extreme Eeconvolution (ED) algorithm, and
is implemented by the function <code>pois_cov_ed</code>. As in the
initialization phase, only data from the genes with strong signals are
used. It should be noted that only data-driven covariance matrices need
to be refined, so they need to be distinguished from the canonical
covariance matrices in the <code>pois_cov_ed</code> call using the
argument <code>ulist.dd</code>. For the purposes of this vignette, we
just ran <code>pois_cov_ed</code> for 10 iterations, but we recommend
performing more iterations when analyzing your own data (the default
number of iterations is 500).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">R</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">X</span><span class="op">)</span>
<span class="co"># Merge data-driven and canonical rank-1 prior covariance matrices.</span>
<span class="va">ulist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res.pca</span><span class="op">$</span><span class="va">ulist</span>,<span class="va">ulist.c</span><span class="op">)</span>
<span class="co"># Specify whether a rank-1 prior covariance matrix is data-driven</span>
<span class="co"># or not (note all full-rank covariance matrices are data-driven</span>
<span class="co"># so do not need this specification).</span>
<span class="va">ulist.dd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">TRUE</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">res.pca</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">R</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> 
<span class="co"># Run ED on the genes with strong signals.</span>
<span class="va">fit.ed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_cov_ed.html">pois_cov_ed</a></span><span class="op">(</span><span class="va">dat</span>,subset <span class="op">=</span> <span class="va">res.pca</span><span class="op">$</span><span class="va">subset</span>,Ulist <span class="op">=</span> <span class="va">res.pca</span><span class="op">$</span><span class="va">Ulist</span>,
                      ulist <span class="op">=</span> <span class="va">ulist</span>,ulist.dd <span class="op">=</span> <span class="va">ulist.dd</span>,ruv <span class="op">=</span> <span class="cn">TRUE</span>,
                      Fuv <span class="op">=</span> <span class="va">Fuv</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span>,init <span class="op">=</span> <span class="va">prefit</span>,
                      control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxiter <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The modification phase adds a small constant <span class="math inline">\(\epsilon^2\)</span> (e.g., 0.01) to the diagonal
entries of the estimated data-driven covariance matrices. This step is
done to alleviate issues of inflated <a href="https://doi.org/10.1093/biostatistics/kxw041" class="external-link">local false sign
rates (lfsr)</a> (the lfsr is analogous to but typically more more
conservative than a local false discovery rate). Similarly, only
data-driven covariance matrices require such modification, and this
information is encoded in the vector <code>epsilon2.G</code> and will be
passed to the function call in the next step.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># add epsilon2*I to each full-rank data-driven prior covariance matrix</span>
<span class="co"># all the full-rank covariance matrices are data-driven and need this modification</span>
<span class="va">Ulist</span> <span class="op">&lt;-</span> <span class="va">fit.ed</span><span class="op">$</span><span class="va">Ulist</span>
<span class="va">H</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Ulist</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">h</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">H</span><span class="op">)</span><span class="op">{</span>
  <span class="va">Uh</span> <span class="op">&lt;-</span> <span class="va">Ulist</span><span class="op">[[</span><span class="va">h</span><span class="op">]</span><span class="op">]</span>
  <span class="va">Uh</span> <span class="op">&lt;-</span> <span class="va">Uh</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">Uh</span><span class="op">)</span><span class="op">)</span>
  <span class="va">Ulist</span><span class="op">[[</span><span class="va">h</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">Uh</span> <span class="op">+</span> <span class="fl">1e-2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">R</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># add epsilon2*I to each rank-1 data-driven prior covariance matrix</span>
<span class="co"># only a subset of the rank-1 covariance matrices are data-driven and need this modification</span>
<span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span>
<span class="va">epsilon2.G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1e-8</span>, <span class="va">G</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">epsilon2.G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span><span class="op">)</span>
<span class="va">epsilon2.G</span><span class="op">[</span><span class="va">ulist.dd</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1e-2</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-3-fit-the-poisson-mash-model">Step 3: Fit the Poisson mash model<a class="anchor" aria-label="anchor" href="#step-3-fit-the-poisson-mash-model"></a>
</h4>
<p>Now we are ready to fit the Poisson mash model to data from all
genes, which is implemented by the function <code>pois_mash</code>.</p>
<p>By default, differences in expression are measured relative to the
mean across all conditions. To change this, see the
<code>pois_mash</code> input arguments <code>C</code> and
<code>res.colnames</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pois_mash.html">pois_mash</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">dat</span>,Ulist <span class="op">=</span> <span class="va">Ulist</span>,ulist <span class="op">=</span> <span class="va">fit.ed</span><span class="op">$</span><span class="va">ulist</span>,
                 ulist.epsilon2 <span class="op">=</span> <span class="va">epsilon2.G</span>,gridmult <span class="op">=</span> <span class="fl">2.5</span>,ruv <span class="op">=</span> <span class="cn">TRUE</span>,
                 Fuv <span class="op">=</span> <span class="va">Fuv</span>,rho <span class="op">=</span> <span class="va">prefit</span><span class="op">$</span><span class="va">rho</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span>,
                 init <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="va">prefit</span><span class="op">$</span><span class="va">mu</span>,psi2 <span class="op">=</span> <span class="va">prefit</span><span class="op">$</span><span class="va">psi2</span><span class="op">)</span>,
                 control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxiter <span class="op">=</span> <span class="fl">10</span>,nc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<p>For this vignette, we ran <code>pois_mash</code> for only 10
iterations, but in practice we recommend setting
<code>maxiter = 100</code>, or perhaps more.</p>
</div>
</div>
<div class="section level2">
<h2 id="examining-the-results">Examining the results<a class="anchor" aria-label="anchor" href="#examining-the-results"></a>
</h2>
<p>We will next look at the results, which are stored in the
<code>res</code> object produced by the call to <code>pois_mash</code>.
Recall that we illustrate how to fit the Poisson mash model using a
subset of the neutrophils data to save time. Here we load in the Poisson
mash fit to the entire neutrophils dataset.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">)</span></code></pre></div>
<p>The output contains posterior summaries of the condition-specific DE
effects relative to the reference condition specified by the user,
including posterior means and posterior standard deviations of the DE
effect sizes (log-fold changes) and measures of significance (lfsr), for
each gene in the dataset. These DE effect posterior summaries are stored
in the <code>result</code> component.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Transform DE effect estimates to the scale of log2 fold change from natural log scale.</span>
<span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">PosteriorMean</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#          Ctrl_2-mean   CCL20-mean    CXCL1-mean    CCL22-mean   CXCL5-mean</span>
<span class="co"># Mrpl15  -0.034042592 -0.007468183  0.0104375255 -0.0212980437 -0.017696220</span>
<span class="co"># Lypla1  -0.009057545 -0.002816239 -0.0312773830 -0.0100671020 -0.011641034</span>
<span class="co"># Tcea1    0.006406279  0.003163708  0.0117792404  0.0112956807  0.007417445</span>
<span class="co"># Atp6v1h  0.005641873  0.004324441 -0.0009984417  0.0001755186  0.002388297</span>
<span class="co"># Rb1cc1   0.005531303  0.010590542 -0.0071790326  0.0060939489  0.003342450</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Obtain gene-specific, condition-specific lfsr for detecting gene-condition DE effects.</span>
<span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">lfsr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#         Ctrl_2-mean CCL20-mean CXCL1-mean CCL22-mean CXCL5-mean</span>
<span class="co"># Mrpl15    0.2527081  0.4012463  0.4501408  0.3236194  0.3458751</span>
<span class="co"># Lypla1    0.4041496  0.4586630  0.2824110  0.3962331  0.3803850</span>
<span class="co"># Tcea1     0.4273815  0.4727043  0.3991172  0.3839887  0.4144509</span>
<span class="co"># Atp6v1h   0.4512667  0.4597141  0.4917574  0.4894398  0.4710016</span>
<span class="co"># Rb1cc1    0.4402427  0.3958559  0.4170690  0.4337283  0.4628668</span></code></pre></div>
<p>We can find the number of DE genes, which are defined as genes having
<em>lfsr</em> less than <span class="math inline">\(\alpha\)</span> in
at least one condition, where <span class="math inline">\(\alpha\)</span> is a threshold specified by the
user (e.g., 0.05).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lfsr</span> <span class="op">&lt;-</span> <span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">lfsr</span>
<span class="co"># Get the indices of identified DE genes.</span>
<span class="va">min.lfsr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">lfsr</span>,<span class="fl">1</span>,<span class="va">min</span><span class="op">)</span>
<span class="co"># Order genes by significance based on mininum lfsr across conditions.</span>
<span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">min.lfsr</span><span class="op">)</span>
<span class="va">lfsr</span> <span class="op">&lt;-</span> <span class="va">lfsr</span><span class="op">[</span><span class="va">rows</span>,<span class="op">]</span>
<span class="va">min.lfsr</span> <span class="op">&lt;-</span> <span class="va">min.lfsr</span><span class="op">[</span><span class="va">rows</span><span class="op">]</span>
<span class="va">de.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">min.lfsr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co"># Get the total number of DE genes.</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">de.genes</span><span class="op">)</span>
<span class="co"># [1] 2636</span></code></pre></div>
<p>The model parameter estimates are stored in the
<code>pois.mash.fit</code> component. In particular, let us look at
<span class="math inline">\(\pi_{kl}\)</span>, the estimates of the
mixture proportions for combinations of different scaling coefficients
<span class="math inline">\(w_l\)</span> and prior covariance matrices
<span class="math inline">\(U_k\)</span>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">pi</span>, cluster_rows<span class="op">=</span><span class="cn">FALSE</span>, cluster_cols<span class="op">=</span><span class="cn">FALSE</span>, fontsize_row<span class="op">=</span><span class="fl">6</span>, fontsize_col<span class="op">=</span><span class="fl">6</span>, main<span class="op">=</span><span class="st">"Proportion estimates of prior covariance matrices"</span><span class="op">)</span></code></pre></div>
<p><img src="intro_poisson_mash_files/figure-html/mixture-proportion-1.png" width="875" style="display: block; margin: auto;"></p>
<p>The data-driven covariance matrix “tPCA” has by far the largest
proportion estimate (91%; summed over all scaling coefficients). We plot
the correlation heatmap for this covariance matrix, which captures the
most frequent pattern of sharing of DE effects across cytokine
treatments in the neutrophils data. For space reason, here we present
the correlation heatmap only for selected conditions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">condition_colors</span><span class="op">)</span>
<span class="va">U</span> <span class="op">&lt;-</span> <span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">Ulist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">U</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">U</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span>
<span class="va">idx.trts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">condition_colors</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">pois.mash.fit</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span>
<span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cov2cor</a></span><span class="op">(</span><span class="va">U</span><span class="op">[</span><span class="va">idx.trts</span>,<span class="va">idx.trts</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">corr</span>,cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,angle_col <span class="op">=</span> <span class="fl">90</span>,
         fontsize <span class="op">=</span> <span class="fl">8</span>,main <span class="op">=</span> <span class="st">"Correlation matrix for\nthe major prior covariance (weight = 91%)"</span><span class="op">)</span></code></pre></div>
<p><img src="intro_poisson_mash_files/figure-html/plot-sharing-pattern-1.png" width="480" style="display: block; margin: auto;"></p>
<p>This correlation heatmap suggests strong sharing of DE effects among
two groups of cytokines (1) IL-1<span class="math inline">\(\alpha\)</span>, IL-1<span class="math inline">\(\beta\)</span>, G-CSF and (2) IL-12 p70, IFN-<span class="math inline">\(\gamma\)</span>, IL-18, IL-15, IL-33. We then plot
the log-fold changes for a few DE genes that either show effect sharing
among (1) or (2), where the reference condition is the mean across all
conditions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### define the color map for LFC</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>,<span class="st">"white"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">99</span><span class="op">)</span>
<span class="va">brks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>, length<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>

<span class="co">### plot a few DE genes that show effect sharing among the first group of cytokines (taken from Figure 7 in our paper)</span>
<span class="va">genes1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Prok2"</span>, <span class="st">"Nos2"</span>, <span class="st">"Teddm3"</span>, <span class="st">"Lipg"</span>, <span class="st">"Saa3"</span>, <span class="st">"Wfdc17"</span>, <span class="st">"Cd38"</span>, <span class="st">"Ggt1"</span>, <span class="st">"Dpep2"</span>, <span class="st">"Mmp19"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">PosteriorMean</span><span class="op">[</span><span class="va">genes1</span>, <span class="va">idx.trts</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="va">cols</span>, breaks <span class="op">=</span> <span class="va">brks</span>, angle_col <span class="op">=</span> <span class="fl">90</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="intro_poisson_mash_files/figure-html/plot-lfc-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">### plot a few DE genes that show effect sharing among the second group of cytokines (taken from Figure 7 in our paper)</span>
<span class="va">genes2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gbp2"</span>, <span class="st">"Gbp3"</span>, <span class="st">"Gbp4"</span>, <span class="st">"Gbp5"</span>, <span class="st">"Ifi47"</span>, <span class="st">"Ifit2"</span>, <span class="st">"Ifit3"</span>, <span class="st">"Socs1"</span>, <span class="st">"Tgtp2"</span>, <span class="st">"Zbp1"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">neutrophils_pois_mash_ruv_fit</span><span class="op">$</span><span class="va">result</span><span class="op">$</span><span class="va">PosteriorMean</span><span class="op">[</span><span class="va">genes2</span>, <span class="va">idx.trts</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="va">cols</span>, breaks <span class="op">=</span> <span class="va">brks</span>, angle_col <span class="op">=</span> <span class="fl">90</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></code></pre></div>
<p><img src="intro_poisson_mash_files/figure-html/plot-lfc-2.png" width="480" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># R version 3.6.2 (2019-12-12)</span>
<span class="co"># Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co"># Running under: macOS Catalina 10.15.7</span>
<span class="co"># </span>
<span class="co"># Matrix products: default</span>
<span class="co"># BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co"># LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co"># </span>
<span class="co"># locale:</span>
<span class="co"># [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co"># </span>
<span class="co"># attached base packages:</span>
<span class="co"># [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co"># [8] methods   base     </span>
<span class="co"># </span>
<span class="co"># other attached packages:</span>
<span class="co">#  [1] poisson.mash.alpha_0.1-98   pheatmap_1.0.12            </span>
<span class="co">#  [3] glmpca_0.2.0                scran_1.14.6               </span>
<span class="co">#  [5] SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.1</span>
<span class="co">#  [7] DelayedArray_0.12.3         BiocParallel_1.18.1        </span>
<span class="co">#  [9] matrixStats_0.63.0          Biobase_2.46.0             </span>
<span class="co"># [11] GenomicRanges_1.38.0        GenomeInfoDb_1.22.1        </span>
<span class="co"># [13] IRanges_2.20.2              S4Vectors_0.24.4           </span>
<span class="co"># [15] BiocGenerics_0.32.0         Matrix_1.3-4               </span>
<span class="co"># </span>
<span class="co"># loaded via a namespace (and not attached):</span>
<span class="co">#  [1] bitops_1.0-6             fs_1.5.2                 RColorBrewer_1.1-2      </span>
<span class="co">#  [4] rprojroot_2.0.3          tools_3.6.2              bslib_0.3.1             </span>
<span class="co">#  [7] utf8_1.1.4               R6_2.4.1                 irlba_2.3.3             </span>
<span class="co"># [10] vipor_0.4.5              DBI_1.1.0                colorspace_1.4-1        </span>
<span class="co"># [13] tidyselect_1.1.1         gridExtra_2.3            compiler_3.6.2          </span>
<span class="co"># [16] cli_3.5.0                BiocNeighbors_1.2.0      desc_1.2.0              </span>
<span class="co"># [19] sass_0.4.0               scales_1.1.0             mvtnorm_1.0-11          </span>
<span class="co"># [22] SQUAREM_2017.10-1        mixsqp_0.3-48            pkgdown_2.0.7           </span>
<span class="co"># [25] systemfonts_1.0.2        stringr_1.4.0            digest_0.6.23           </span>
<span class="co"># [28] rmarkdown_2.21           XVector_0.26.0           scater_1.14.6           </span>
<span class="co"># [31] pkgconfig_2.0.3          htmltools_0.5.4          highr_0.8               </span>
<span class="co"># [34] invgamma_1.1             fastmap_1.1.0            limma_3.42.2            </span>
<span class="co"># [37] rlang_1.0.6              DelayedMatrixStats_1.6.1 jquerylib_0.1.4         </span>
<span class="co"># [40] generics_0.0.2           jsonlite_1.7.2           dplyr_1.0.7             </span>
<span class="co"># [43] RCurl_1.98-1.2           magrittr_2.0.1           BiocSingular_1.2.2      </span>
<span class="co"># [46] GenomeInfoDbData_1.2.2   Rcpp_1.0.8               ggbeeswarm_0.6.0        </span>
<span class="co"># [49] munsell_0.5.0            fansi_0.4.0              abind_1.4-5             </span>
<span class="co"># [52] viridis_0.5.1            lifecycle_1.0.3          stringi_1.4.3           </span>
<span class="co"># [55] yaml_2.2.0               edgeR_3.28.1             MASS_7.3-51.4           </span>
<span class="co"># [58] zlibbioc_1.32.0          grid_3.6.2               dqrng_0.2.1             </span>
<span class="co"># [61] crayon_1.4.1             lattice_0.20-38          locfit_1.5-9.4          </span>
<span class="co"># [64] knitr_1.37               pillar_1.6.2             igraph_1.2.5            </span>
<span class="co"># [67] glue_1.4.2               evaluate_0.14            vctrs_0.3.8             </span>
<span class="co"># [70] gtable_0.3.0             purrr_0.3.4              assertthat_0.2.1        </span>
<span class="co"># [73] ashr_2.2-57              ggplot2_3.3.6            xfun_0.36               </span>
<span class="co"># [76] rsvd_1.0.2               ragg_0.3.1               viridisLite_0.3.0       </span>
<span class="co"># [79] truncnorm_1.0-8          tibble_3.1.3             poilog_0.4              </span>
<span class="co"># [82] beeswarm_0.2.3           memoise_1.1.0            seqgendiff_1.2.2        </span>
<span class="co"># [85] statmod_1.4.34           ellipsis_0.3.2</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yusha Liu, Peter Carbonetto, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
